src/common/cockpitunicode.c

```diff
@@ -21,6 +21,27 @@
 
 #include "cockpitunicode.h"
 
+gboolean
+cockpit_unicode_has_incomplete_ending (GBytes *input)
+{
+  const gchar *data;
+  const gchar *end;
+  gsize length;
+
+  data = g_bytes_get_data (input, &length);
+  if (g_utf8_validate (data, length, &end))
+    return FALSE;
+
+  do
+    {
+      length -= (end - data) + 1;
+      data = end + 1;
+    }
+  while (!g_utf8_validate (data, length, &end));
```

Comment:
This seems to ignore  invalid UTF-8 blocks in the *middle* of the string and then keep trying to find some valid UTF-8 again. That seems a bit over-opportunistic at first sight - is there an use case where the reader can run into dangling UTF-8 endings with the whole string *not* being UTF-8?

For my own notekeeping to understand the algorithm: After the first validate, where "AAA" and BBB are good UTF-8, "xy" is an invalid block, it looks like:
```
validate("AAAxyBBB", length=8) == false
             ⬑ end
```
i. e. `end` points to the `x`; due to the +1 this gets skipped and after the first `do` loop the loop condition test is
```
validate("yBBB", length=4) == false
          ⬑ end
```

As assumed, `y` is invalid, so  supposedly `end` points to the `y`, and length and data are advanced by one. Next condition test is
```
validate("BBB", length=3) == true
             ⬑ end
```
i. e. `end` now points to the NUL terminator as the whole remainder of the string is valid. But `length` is still 3, so function returns FALSE, as expected.

If OTOH we only start with `AAAxy`, then after the first step we end up at
```
validate("y", length=1) == false
          ⬑ end
```
and length gets reduced to 0, thus the function returns TRUE. This is also as expected.

Reply:
Peter clarified that in the meeting: There are situations where we *do* have invalid UTF-8 in the middle, as we read from all sorts of sources (journal, processes, files, etc.), which we'll later replace with Unicode "unknown character" symbols. So we really just need to avoid invalid UTF-8 at the end. So this makes perfect sense, thanks!